### ISSUE TITLE
`<atomic>`: Avoid checks for `cmpxchg16b` availability

### ISSUE THREAD

**Issue by AlexGuteniev** [microsoft/STL#4481]
Ideally, we'd want drop it completely, but need to confirm if these CPUs are not needed to be supported. They can run on systems before Windows 8.1

Alternatively we may have CPU assumption based on a certain `/arch` mode. I don't know which, though. I'm sure AVX512 implies that, but looks like the bar can be lower.

**Comment by Alcaro** [comment]
Might as well repost from Discord.

According to [assorted](https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html) [sources](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=107337), every currently-existing CPU that supports AVX also supports CX16, with exception of VIA eden-x4. If ignoring a few other VIAs, everything with SSSE3 has CX16. It's also present in -march=x86-64-v2, and mandatory for w8+, so future CPUs without that instruction seem quite unlikely to me.

I'd say the first step should be to ask the VIA folks if gcc's lack of that instruction is a gcc bug - [this guy](https://discord.com/channels/737189251069771789/737734473751330856/1181320524178149427) says AVX-enabled VIAs have atomic xmm moves (though I wasn't able to find that llvm bug), and if that one's atomic, not supporting CX16 feels like a quite strange move to me.

If the instruction is absent, and supporting that CPU is deemed valuable, then the only /arch that guarantees CX16 is avx512, which is (to my knowledge) a quite rare value, hardly worth optimizing for. Perhaps the best solution is simply wait until w7 departs.

(Probably not far off. Rest in peace.)

**Comment by StephanTLavavej** [comment]
At this time, we *definitely* still need to support ancient CPUs, e.g. those that don't have SSE4.2. (There was a recent compiler back-end bug regarding unguarded use of the `popcnt` instruction, which wrought havoc before it was fixed.)

I'm also not exactly sure of the precise timing of when this instruction appeared (with respect to increasing ISA feature sets) and when we'll be able to start assuming its presence. We talked about this at the weekly maintainer meeting, and this really seems like a vNext optimization, creating a clean baseline for performance going forward.
